engine/engine.go

```diff
@@ -152,28 +154,48 @@ func DownloadEngine(ctx context.Context, opts *options.TerragruntOptions) error
 		return nil
 	}
 	downloadFile := filepath.Join(path, enginePackageName(e))
-	var downloadURL string
+
+	downloads := make(map[string]string)
+	checksumFile := ""
+	checksumSigFile := ""
 	if strings.HasPrefix(e.Source, "http") {
 		// if source starts with absolute path, download as is
-		downloadURL = e.Source
+		downloads[e.Source] = downloadFile
 	} else {
-		// Archive support documented in https://github.com/hashicorp/go-getter?tab=readme-ov-file#unarchiving
-		downloadURL = fmt.Sprintf("https://%s/releases/download/%s/%s",
-			e.Source, e.Version, enginePackageName(e))
-	}
+		baseURL := fmt.Sprintf("https://%s/releases/download/%s", e.Source, e.Version)
+
+		// URLs and their corresponding local paths
+		checksumFile = filepath.Join(path, engineChecksumName(e))
+		checksumSigFile = filepath.Join(path, engineChecksumSigName(e))
+		downloads[fmt.Sprintf("%s/%s", baseURL, enginePackageName(e))] = downloadFile
+		downloads[fmt.Sprintf("%s/%s", baseURL, engineChecksumName(e))] = checksumFile
+		downloads[fmt.Sprintf("%s/%s.sig", baseURL, engineChecksumName(e))] = checksumSigFile
+	}
+
+	for url, path := range downloads {
+		opts.Logger.Infof("Downloading %s to %s", url, path)
+		client := &getter.Client{
+			Ctx:           ctx,
+			Src:           url,
+			Dst:           path,
+			Mode:          getter.ClientModeFile,
+			Decompressors: map[string]getter.Decompressor{},
+		}
 
-	client := &getter.Client{
-		Ctx:           ctx,
-		Src:           downloadURL,
-		Dst:           downloadFile,
-		Mode:          getter.ClientModeFile,
-		Decompressors: map[string]getter.Decompressor{},
+		if err := client.Get(); err != nil {
+			return errors.WithStackTrace(err)
+		}
 	}
 
-	if err := client.Get(); err != nil {
-		return errors.WithStackTrace(err)
+	if !skipEngineCheck() && checksumFile != "" && checksumSigFile != "" {
+		opts.Logger.Infof("Verifying checksum for %s", downloadFile)
+		if err := verifyFile(downloadFile, checksumFile, checksumSigFile); err != nil {
+			return errors.WithStackTrace(err)
+		}
+	} else {
+		opts.Logger.Warnf("Skipping verification for %s", downloadFile)
 	}
-	opts.Logger.Infof("Engine downloaded to %s", downloadFile)
+
 	if err := extractArchive(opts, downloadFile, localEngineFile); err != nil {
```

Comment:
I know that this was added before, but consider to use:

```go
// Borrow the "unpack a zip cache into a target directory" logic from go-getter
var unzip = getter.ZipDecompressor{}

func(.... 
	if err := unzip.Decompress(dst, src, isDir, umask ); err != nil {
		return errors.WithStackTrace(err)
	}
```

Reply:
Hm, I think `archiver` package has more applications here:
* it can detect if the downloaded file is an archive or not - people may use engine files as is
* I think it can also extract more archive formats in case if people will use different archiving tools

Reply:
Roger